### DISCUSSION TITLE
`show_if` variation

### DISCUSSION THREAD

**Discussion by willking98** [surveydown-dev/surveydown#98] [Q&A] [answered]
### Description

I think the `show_if` function is working slightly differently when used in conjunction with `sd_question` and `sd_output`. For me, I have to click next in order to render the conditional question when using `sd_output` whereas when using `sd_question` the conditional question pops up instantly when the condition is satisfied (once correct option has been selected). I think this might be a speed issue as sometimes it renders and other times I have to click next.

**Comment by jhelvy** [comment]
Can you provide a quick demo? like a simple `survey.qmd` and `app.R` where this happens?

**Comment by jhelvy** [comment]
Okay I can reproduce this. In this example the hidden question won't show at all when you choose "other" on the first question:

`survey.qmd`:

````qmd
---
format: html
echo: false
warning: false
---

```{r}
library(surveydown)
```

::: {#page1 .sd-page}

```{r}
sd_question(
  type  = 'mc',
  id    = 'penguins',
  label = "Which is your favorite type of penguin?",
  option = c(
    'Adélie'    = 'adelie',
    'Chinstrap' = 'chinstrap',
    'Gentoo'    = 'gentoo',
    'Other'     = 'other'
  )
)

sd_output(id = 'penguins_other', type = 'question')

sd_next()
```

:::

::: {#end .sd-page}

## This is the end of the demo survey showcasing the conditional display feature.

```{r}
sd_close()
```

:::

````


`app.R`:

```r
library(surveydown)

server <- function(input, output, session) {

  sd_question(
    type  = "text",
    id    = "penguins_other",
    label = "Please specify the other penguin type:"
  )

  sd_show_if(
    input$penguins == "other" ~ "penguins_other"
  )

  sd_server()

}

shiny::shinyApp(ui = sd_ui(), server = server)

```

**Reply by willking98** [reply]
I think my conditional questions are showing because of `all_questions_required = TRUE`. You should be able to see what I mean by running the survey I have developed [here](https://github.com/willking98/demo-choice-based-conjoint)

**Comment by jhelvy** [comment]
Okay I just pushed a patch. Try reinstalling and see if this works:

```r
remotes::install_github("surveydown-dev/surveydown", force = TRUE)
```

**Reply by willking98** [reply]
Works perfectly, nice work. Cheers!

**Comment by jhelvy** [comment]
Such a small little bug. Thanks for the quick notice. We tried to make sure everything was still working after making such big changes to the architecture. One more thing we noticed is the little red asterisk for required questions won't show on questions defined in the  server. The required feature still works, but the question label is just missing the red asterisk. We'll have it patched soon.

**Reply by willking98** [reply]
No problem at all. Thanks for getting it sorted so quickly. 

The other thing I wanted to mention which seems to have changed in this update is the formatting of `mc_buttons` see images below. Is it still possible to use the old formatting? Also is it possible to force each option to appear on a new line? 

**New**
![image](https://github.com/user-attachments/assets/36d17a5f-507d-40ed-af3f-601320422152)

**Old**
![image](https://github.com/user-attachments/assets/5cb24ac4-6cd1-4f6b-8890-42bd599d0649)

**Comment by jhelvy** [comment]
Ah, good catch. @pingfan-hu can you take a look at this?

By moving from a Quarto shiny document to now just a regular Quarto doc + a shiny app, we lost some of the nice formatting that Quarto handled for us. Pingfan has been working on the UI to try and make things looks similar to how they rendered before.

**Comment by pingfan-hu** [comment]
I didn't intentionally modify the buttons. The old layout has equal width because quarto renders it this way. Now since it becomes a Shiny app, it's rendered differently. I'll see if I can do something in the css or insert some JS codes.

**Comment by pingfan-hu** [comment, marked as answer]
Hi @willking98 , I've solved your problem. You can see the screenshot here:
<img width="600" alt="image" src="https://github.com/user-attachments/assets/1cad13f3-981a-427b-a122-2b5f997cf1f2">

This is done by defining the `direction` parameter as `"vertical"` in `mc_buttons`. This parameter is available for both `mc_buttons` and `mc_multiple_buttons`, default to `"horizontal"`.

For your convenience, I've modified all your `mc_buttons` questions to have vertical alignment:
[demo-choice-based-conjoint-modified.zip](https://github.com/user-attachments/files/17121768/demo-choice-based-conjoint-modified.zip)

Another parameter is called `"width"`. You can define it as `"400px"` or `"80%"` so that you can control the width of the question container, which then affects the box length. The texts will automatically wrap if they are too long, like this:
<img width="600" alt="image" src="https://github.com/user-attachments/assets/e4d5f5b9-9f3b-46eb-b11c-f8b4ffc242d3">


Please update our package and try it:
```r
remotes::install_github("surveydown-dev/surveydown", force = TRUE)
```

Now we have a bug that if you define width as 80%, the question container will take 80% of the page width, but then the button width will be 80% of the container, making the button width being 64% of the page. I will fix it later.

**Comment by pingfan-hu** [comment]
@willking98 As a follow-up, the `width` problem has been solved. Now if you define a `width` value to your questions, say `"80%"`, `"60%"`, etc., only the question container will shrink, but the buttons will still fully span.

**Comment by willking98** [comment]
Brilliant, that now works really well. Thanks @pingfan-hu!
